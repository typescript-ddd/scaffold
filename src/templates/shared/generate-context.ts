import { SourceFile } from "ts-morph";

export type ContextImportDeclaration = { moduleSpecifier: string; namedImports: string[] };

export interface GenerateContext {
    resolveDir(type: "domain" | "infrastructure" | "application" | "presentation", subPath?: string): string;
    fileName: string;
    addImportDeclaration(declaration: ContextImportDeclaration): void;
    getImportDeclarations(moduleSpecifier?: string): ContextImportDeclaration[];
    getModuleSpecifiers(): string[];
    getNamedImports(moduleSpecifier?: string): string[];    
    writeSourceFileImports: (sourceFile: SourceFile) => void;
    clearImportDeclarations(): void;
    tokenize: (text: string, tokens: { name: string }[]) => string;
}