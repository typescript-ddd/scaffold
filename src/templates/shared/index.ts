export * from "./context-builder";
export * from "./generate-context";
export * from "./template";