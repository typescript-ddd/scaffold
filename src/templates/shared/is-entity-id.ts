export const isEntityId = (str: string, entity: string) =>
  `${entity}Id` === str;
