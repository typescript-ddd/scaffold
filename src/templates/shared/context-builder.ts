import { GenerateContext, ContextImportDeclaration } from "./generate-context";
import { arrays } from "../../utils";
import { SourceFile } from "ts-morph";

export class ContextBuilder implements GenerateContext {
  private _fileName: string;
  private _importDeclarations: ContextImportDeclaration[] = [];

  private constructor(
    private rootDir: string = "@/src",
    fileName: string = "example.ts"
  ) {
    this._fileName = fileName;
  }

  resolveDir(
    type: "domain" | "infrastructure" | "application" | "presentation",
    subPath?: string
  ): string {
    switch (type) {
      case "domain":
        return `${this.rootDir}/domain${subPath ? `/${subPath}` : ""}`;
      case "infrastructure":
        return `${this.rootDir}/infrastructure${subPath ? `/${subPath}` : ""}`;
      case "application":
        return `${this.rootDir}/application${subPath ? `/${subPath}` : ""}`;
      case "presentation":
        return `${this.rootDir}/presentation${subPath ? `/${subPath}` : ""}`;
      default:
        throw new Error(`Unknown type ${type} for resolveDir`);
    }
  }

  getImportDeclarations(moduleSpecifier?: string): ContextImportDeclaration[] {
    const modules = this.getModuleSpecifiers();

    const declarations: ContextImportDeclaration[] = [];

    for (const currentModule of modules) {
      const namedImports = this.getNamedImports(currentModule);
      declarations.push({ moduleSpecifier: currentModule, namedImports });
    }

    if (moduleSpecifier) {
      return declarations.filter((x) => x.moduleSpecifier === moduleSpecifier);
    }

    return declarations;
  }

  addImportDeclaration(declaration: ContextImportDeclaration): void {
    this._importDeclarations.push(declaration);
  }

  getModuleSpecifiers(): string[] {
    return arrays.distinct(
      this._importDeclarations.map((x) => x.moduleSpecifier)
    );
  }

  getNamedImports(moduleSpecifier: string): string[] {
    return arrays.distinct(
      this._importDeclarations
        .filter((x) => x.moduleSpecifier === moduleSpecifier)
        .map((x) => x.namedImports)
        .flat()
    );
  }

  writeSourceFileImports(sourceFile: SourceFile) {
    const importDeclarations = this.getImportDeclarations();
    const moduleSpecifiers = this.getModuleSpecifiers();

    const coreImports = importDeclarations.filter(
      (x) => x.moduleSpecifier === "@typescript-ddd/core"
    );

    if (coreImports) {
      sourceFile.addImportDeclaration({
        moduleSpecifier: `@typescript-ddd/core`,
        namedImports: coreImports.map((x) => x.namedImports).flat(),
      });
    }

    // Non core and root imports
    for (const moduleSpecifier of moduleSpecifiers
      .filter(
        (x) => x !== "@typescript-ddd/core" && !x.startsWith(this.rootDir)
      )
      .sort((a, b) => {
        if (a < b) return -1;
        if (a > b) return 1;
        return 0;
      })) {
      sourceFile.addImportDeclaration({
        moduleSpecifier,
        namedImports: this.getNamedImports(moduleSpecifier),
      });
    }

    // Root imports
    for (const moduleSpecifier of moduleSpecifiers
      .filter((x) => x.startsWith(this.rootDir))
      .sort((a, b) => {
        if (a < b) return -1;
        if (a > b) return 1;
        return 0;
      })) {
      sourceFile.addImportDeclaration({
        moduleSpecifier,
        namedImports: this.getNamedImports(moduleSpecifier),
      });
    }
  }

  clearImportDeclarations(): void {
    this._importDeclarations = [];
  }

  tokenize(text: string, tokens: { name: string }[]): string {
    let output = text;
    tokens.forEach((token) => {
      output = output.replace(`{{${token.name}}}`, "${" + token.name + "}");
    });
    return "`" + output + "`";
  }

  get fileName(): string {
    return this._fileName;
  }

  static build(rootDir: string): ContextBuilder {
    return new ContextBuilder(rootDir);
  }
}
