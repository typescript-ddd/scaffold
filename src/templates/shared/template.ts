import { GenerateContext } from "./generate-context";

export type TemplateEnum = Record<string, string>;
export type TemplatePropertyValueType =
  | "string"
  | "number"
  | "boolean"
  | "date";

export type TemplateEnumOption = {
  name: string;
  label: string;
  description: string;
  enum: TemplateEnum;
  optionType: "enum";
};

export type TemplateStringOption = {
  name: string;
  label: string;
  description: string;
  optionType: "string";
};

export type TemplateTokenStringOption = {
  name: string;
  label: string;
  description: string;
  optionType: "tokenString";
};

export type TemplateBooleanOption = {
  name: string;
  label: string;
  description: string;
  optionType: "boolean";
};

export type TemplateNumberOption = {
  name: string;
  label: string;
  description: string;
  optionType: "number";
};

export type TemplateDateOption = {
  name: string;
  label: string;
  description: string;
  optionType: "date";
};

export type TemplatePropertyOption = {
  name: string;
  valueType: TemplatePropertyValueType;
  nullable?: boolean;
  prop?: string;
};

export type TemplatePropertyValue = {
  name: string;
  valueType: string;
  nullable?: boolean;
  prop?: string;
};

export type TemplatePropertiesOption = {
  name: string;
  label: string;
  description: string;
  optionType: "properties";
};

export type TemplateOption =
  | TemplateEnumOption
  | TemplateStringOption
  | TemplateBooleanOption
  | TemplateNumberOption
  | TemplateDateOption
  | TemplatePropertiesOption
  | TemplateTokenStringOption;

export type Maybe<T> = T | null | undefined;

export interface Template<
  TValues extends { [K in keyof TOptions]: any},
  TOptions extends { [key: string]: TemplateOption }
> {
  name: string;
  description: string;
  options: TOptions;
  defaultValues: TValues;
  generate(values: TValues, context: GenerateContext): string;
}