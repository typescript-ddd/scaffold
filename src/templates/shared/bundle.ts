import { GenerateContext, Chunk } from "./generate-context";
import { Template, TemplateOption } from "./template";

export interface Bundle<
  TValues extends { [K in keyof TOptions]: any },
  TOptions extends { [key: string]: TemplateOption }
> extends Omit<Template<TValues, TOptions>, "generate"> {
  generate(value: TValues, context: GenerateContext): Chunk[];
}
