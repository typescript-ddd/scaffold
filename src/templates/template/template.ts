import { Project, Scope } from "ts-morph";
import { strings } from "../../utils";
import { GenerateContext } from "../shared";

export type GenerateTemplateOptions = {
  name: string;
};

export const generateTemplate = (
  options: GenerateTemplateOptions,
  context: GenerateContext
) => {
  const { name } = options;
  const className = strings.capitalize(name);

  const project = new Project();
  const sourceFile = project.createSourceFile(
    `${context.fileName}.ts`,
    `
export class ${className} {}    
`
  );

  context.addImportDeclaration({
    moduleSpecifier: `@typescript-ddd/core`,
    namedImports: ["Entity"],
  });

  context.writeSourceFileImports(sourceFile);

  const classDeclaration = sourceFile.getClass(className);
  classDeclaration.addJsDoc({
    description: `${strings.describe(name)}.`,
  });

  const constructorDeclaration = classDeclaration.addConstructor({
    scope: Scope.Private,
    parameters: [{ name: "value", type: "string" }],
  });
  constructorDeclaration.setBodyText(`super(value);`);
  constructorDeclaration.addJsDoc({
    description: `Initializes a new instance of ${strings.describe(
      name
    )} identifier.`,
    tags: [{ tagName: "param", text: "value The value of the identifier." }],
  });

  let output = sourceFile.getFullText();

  return output;
};
