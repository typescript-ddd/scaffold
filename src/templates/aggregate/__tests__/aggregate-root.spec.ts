import { ContextBuilder, GenerateContext } from "../../shared";
import { AggregateRootTemplate } from "../aggregate-root.template";

describe("Aggregate", () => {
  let context: GenerateContext;
  let template: AggregateRootTemplate;

  beforeEach(() => {
    context = ContextBuilder.build("@src");
    template = new AggregateRootTemplate();
  });

  it("Should render the aggregate root component", () => {
    
    context.addImportDeclaration({
      moduleSpecifier: "@src/domain/models",
      namedImports: ["Phone"],
    });

    const output = template.generate(
      {
        entityName: "User",
        properties: [{ name: "name", valueType: "string" }, { name: "phone", valueType: "Phone"}],
        trackable: false,
      },
      context
    );

    expect(output).toMatchSnapshot();
  });

  it("Should render the trackable aggregate root component", () => {
    const output = template.generate(
      {
        entityName: "User",
        properties: [{ name: "name", valueType: "string" }],
        trackable: true,
      },
      context
    );

    expect(output).toMatchSnapshot();
  });
});
