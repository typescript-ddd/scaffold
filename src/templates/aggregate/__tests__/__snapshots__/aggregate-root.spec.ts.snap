// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Aggregate Should render the aggregate root component 1`] = `
"import { AggregateRoot, EntityCreateProps, EntityUpdateProps  } from "@typescript-ddd/core";
import { UserCreatedEvent, UserUpdatedEvent, UserDeletedEvent } from "@src/domain/events";
import { Phone, UserId } from "@src/domain/models";

/**
 * Represents the properties required to create a user.
 * @extends {EntityCreateProps}
 */
export interface CreateUserProps extends EntityCreateProps {
    name: string;
    phone: Phone;
}
/**
 * Represents the properties required to update a user.
 * @extends {EntityUpdateProps}
 */
export interface UpdateUserProps extends EntityUpdateProps<UserId> {
    name: string;
    phone: Phone;
}
/**
 * Represents a user aggregate.
 * @extends AggregateRoot
 */
export class User extends AggregateRoot<UserId> {
    /**
     * Initializes a new instance of a user.
     * @param {UserId} id - An optional identifier of the user.
     * @param {string} name - The name of the user.
     * @param {Phone} phone - The phone of the user.
     */
    private constructor(private name: string, private phone: Phone, id?: UserId) {
        super(id ?? UserId.create());
        if (!id) {
          this.applyChange(new UserCreatedEvent(this.id, this));
        }
    }

    /**
     * Gets the name of the user.
     * @returns {string} - The name of the user.
     */
    getName(): string {
        return this.name;
    }

    /**
     * Gets the phone of the user.
     * @returns {Phone} - The phone of the user.
     */
    getPhone(): Phone {
        return this.phone;
    }

    /**
     * Update the user.
     * @param {UpdateUserProps} props - The properties to update.
     * @returns void
     */
    update(props: UpdateUserProps): void {
        // TODO: validate props
        this.name = props.name;
        this.phone = props.phone;
        this.applyChange(new UserUpdatedEvent(this.id, this));
    }

    delete(): void {
        this.applyChange(new UserDeletedEvent(this.id, this));
    }

    static create(props: CreateUserProps, id?: UserId): User {
        const user = new User(
          props.name,
          props.phone,
          id
        );
        return user;
    }
}    
"
`;

exports[`Aggregate Should render the trackable aggregate root component 1`] = `
"import { TrackableAggregateRoot, EntityCreateProps, EntityUpdateProps, UtcDate } from "@typescript-ddd/core";
import { UserCreatedEvent, UserUpdatedEvent, UserDeletedEvent } from "@src/domain/events";
import { UserId } from "@src/domain/models";

/**
 * Represents the properties required to create a user.
 * @extends {EntityCreateProps}
 */
export interface CreateUserProps extends EntityCreateProps {
    name: string;
    createdAtUtc?: UtcDate;
    updatedAtUtc?: UtcDate;
}
/**
 * Represents the properties required to update a user.
 * @extends {EntityUpdateProps}
 */
export interface UpdateUserProps extends EntityUpdateProps<UserId> {
    name: string;
}
/**
 * Represents a user aggregate.
 * @extends TrackableAggregateRoot
 */
export class User extends TrackableAggregateRoot<UserId> {
    /**
     * Initializes a new instance of a user.
     * @param {UserId} id - An optional identifier of the user.
     * @param {string} name - The name of the user.
     * @param {UtcDate} createdAtUtc - An optional creation date.
     * @param {UtcDate} updatedAtUtc - An optional last update date.
     */
    private constructor(private name: string, createdAtUtc?: UtcDate, updatedAtUtc?: UtcDate, id?: UserId) {
        super(id ?? UserId.create(), createdAtUtc, updatedAtUtc);
        if (!id) {
          this.applyChange(new UserCreatedEvent(this.id, this));
        }
    }

    /**
     * Gets the name of the user.
     * @returns {string} - The name of the user.
     */
    getName(): string {
        return this.name;
    }

    /**
     * Update the user.
     * @param {UpdateUserProps} props - The properties to update.
     * @returns void
     */
    update(props: UpdateUserProps): void {
        // TODO: validate props
        this.name = props.name;
        this.updatedAtUtc = UtcDate.now();
        this.applyChange(new UserUpdatedEvent(this.id, this));
    }

    delete(): void {
        this.applyChange(new UserDeletedEvent(this.id, this));
    }

    static create(props: CreateUserProps, id?: UserId): User {
        const user = new User(
          props.name,
          id
        );
        return user;
    }
}    
"
`;
